Sources/SIL/SILAnalysis.swift

```diff
@@ -195,11 +196,193 @@ extension Array: AlphaConvertible where Element: AlphaConvertible {
+    case let .genericType(parameters, requirements, subtype):
+        return .genericType(
+            parameters,
+            requirements,
+            substitute(subtype, using: { parameters.contains($0) ? .namedType($0) : s($0) }))
+    case let .namedType(name):
+        return s(name)
+    case let .selectType(subtype, name):
+        return .selectType(substitute(subtype, using: s), name)
+    case .selfType:
+        return .selfType
+    case let .specializedType(genericType, arguments):
+        return .specializedType(
+            substitute(genericType, using: s), arguments.map { substitute($0, using: s) })
+    case let .tupleType(elementTypes):
+        return .tupleType(elementTypes.map { substitute($0, using: s) })
+    case let .withOwnership(attribute, subtype):
+        return .withOwnership(attribute, substitute(subtype, using: s))
+    }
+}
+
+func specialize(_ type: Type, to arguments: [Type]) -> Type {
+    switch type {
+    case let .addressType(subtype):
+        return .addressType(specialize(subtype, to: arguments))
+    case let .attributedType(attributes, subtype):
+        return .attributedType(attributes, specialize(subtype, to: arguments))
+    case let .genericType(parameters, _, subtype):
+        guard parameters.count == arguments.count else {
+            fatalError(
+                "Specializing a generic type with \(parameters.count) parameters using \(arguments.count) arguments"
+            )
+        }
+        let valuation = [String: Type](
+            zip(parameters, arguments),
+            uniquingKeysWith: { _, _ in fatalError("Duplicate parameter names in generic type") })
+        return substitute(subtype, using: { valuation[$0]! })
+    case let .selectType(subtype, name):
+        return .selectType(specialize(subtype, to: arguments), name)
+    case let .withOwnership(attribute, subtype):
+        return .withOwnership(attribute, specialize(subtype, to: arguments))
+    case .coroutineTokenType: fallthrough
+    case .functionType(_, _): fallthrough
+    case .namedType(_): fallthrough
+    case .selfType: fallthrough
+    case .specializedType(_, _): fallthrough
+    case .tupleType(_):
+        fatalError("Specializing a type that is not generic")
+    }
+}
+
+func destructFunctionType(_ type: Type) -> (arguments: [Type], result: Type) {
+    switch type {
+    case let .attributedType(_, subtype):
+        return destructFunctionType(subtype)
+    case let .functionType(arguments, result):
+        return (arguments, result)
+    case let .genericType(_, _, subtype):
+        return destructFunctionType(subtype)
+    case let .withOwnership(_, subtype):
+        return destructFunctionType(subtype)
+    case .addressType(_): fallthrough
+    case .coroutineTokenType: fallthrough
+    case .namedType(_): fallthrough
+    case .selectType(_, _): fallthrough
+    case .selfType: fallthrough
+    case .specializedType(_, _): fallthrough
+    case .tupleType(_):
```

Comment:
I think you can do something like `case .foo(_), .bar(_, _): ...`

Reply:
I remember trying this once, and it didn't work, so my conclusion was that this chaining is limited to e.g. integer literals, but I might be wrong.